engines/grim/hotspot.cpp

```diff
@@ -0,0 +1,768 @@
+	update();
+}
+
+void HotspotMan::drawActive() {
+	if (_flashHS) {
+		unsigned int curTime = g_system->getMillis();
+		unsigned int delta = curTime - _flashStart;
+		if (delta >= 1700) _flashHS = false;
+		if ((delta % 900) < 450) {
+			int setup = g_grim->getCurrSet()->getSetup();
+			Common::Array<Hotspot> &hotspots = _hotspots[activeSet()];
+			for (size_t i = 0, total = hotspots.size(); i < total; ++i) {
+				Hotspot &hs = hotspots[i];
+				if (hs._setup == setup && (hs._objId<=0 || _hotobject[hs._objId]._active)) {
+					if ((_ctrlMode == Dialog && hs._type >= 20) ||
+						(_ctrlMode == Special && hs._type >= 10 && hs._type < 20) ||
+						(_ctrlMode == Linear && hs._type >= 10 && hs._type < 20) ||
+						((_ctrlMode == Normal || _ctrlMode == NoWalk) && hs._type < 10)) {
+						int cu = hs._type % 10;
+						if (hs._type == 3) cu = 2;
+						if (hs._type == 4) cu = 1;
+						Common::Point p = hs._region.center();
+						_flashBitmaps[cu]->draw(p.x - 15, p.y - 15);
+					}
+				}
+			}
+		}
+	}
+
+	if (_ctrlMode == Inventory) {
+		const int dx = 4;
+		const int numRows = (_inventory.size() - 1) / _cols + 1;
+		int x1 = _x0 + _w * _cols, y1 = _y0 + _h * numRows;
+		g_driver->blackbox(0, 0, 640, 480, 0.4);
+		g_driver->blackbox(_x0 - dx, _y0 - dx, x1 + dx, y1 + dx, 0.9);
+		PrimitiveObject rect;
+		Color col(195, 195, 100);
+		rect.createRectangle(Common::Point(_x0 - dx, _y0 - dx), Common::Point(x1 + dx, y1 + dx), col, false);
+		rect.draw();
+
+		for (int j = 0, idx = 0; j < _rows; j++) {
+			for (int i = 0; i < _cols; ++i, ++idx) {
+				if (idx >= (int)_inventory.size())
+					break;
+				_inventory[idx]._bmp->draw(_x0 + i * _w, _y0 + j * _h);
+			}
+		}
+	}
+}
+
+void HotspotMan::updatePerspective() {
+	g_grim->getCurrSet()->setupCamera();
+	for (size_t i = 0; i < _hotobject.size(); ++i) {
+		int x = 0, y = 0;
+		g_driver->worldToScreen(_hotobject[i]._pos, x, y);
+		_hotobject[i]._rect = Common::Rect(x - 10, y - 10, x + 10, y + 10);
+	}
+	restoreCursor();
+}
+
+Common::String HotspotMan::activeSet() {
+	return (_curOption.empty() || _ctrlMode != Options) ?
+				g_grim->getCurrSet()->getName() : _curOption;
+}
+
+// get min distance s for x0 + s * x1, y0 + t * y1
+static double lineLineDist(
+		const Math::Vector3d &x0, const Math::Vector3d &x1,
+		const Math::Vector3d &y0, const Math::Vector3d &y1) {
+	double a = x1.dotProduct(x1), b = -x1.dotProduct(y1), c = y1.dotProduct(y1);
+	double d = -x1.dotProduct(x0 - y0), e = y1.dotProduct(x0 - y0);
+	return (c * d - b * e) / (c * a - b * b);
+}
+
+void HotspotMan::event(const Common::Point &cursor, const Common::Event &ev, bool doubleClick) {
+	bool climbing = LuaBase::instance()->queryVariable("system.currentActor.is_climbing", false) != 0;
+
+	int button = 0;
+	if (ev.type == Common::EVENT_LBUTTONDOWN)
+		button = 1;
+	else if (ev.type == Common::EVENT_RBUTTONDOWN)
+		button = 2;
+	else if (ev.type == Common::EVENT_MBUTTONDOWN)
+		button = 3;
+	if (ev.kbd.hasFlags(Common::KBD_ALT))
```

Comment:
why we add ATL and CTRL?

Reply:
Same big commit :/

Reply:
I'll look into

Reply:
can you tell me for what ALT and CTRL are used here? they are pushed into Lua some unknown reasons. since I don't see Lua script source, hard to guess

Reply:
This is the lua bytecode.
[mouse.zip](https://github.com/scummvm/scummvm/files/7464089/mouse.zip)

I can't decompile it. :/

Reply:
you can use 'delua' from residualvm-tools, tools from scummvm-tools are not yet enabled.
I briefly looked into, and things are started to worry me. that patch can not have any original scripts code. it needs to be reviewed :/ but later

Reply:
Awesome!

This is the source file: [mouse.zip](https://github.com/scummvm/scummvm/files/7464151/mouse.zip)